cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll

```diff
@@ -175,6 +176,23 @@ private predicate nodeIsBarrier(DataFlow::Node node) {
     readsVariable(node.asInstruction(), checkedVar) and
     hasUpperBoundsCheck(checkedVar)
   )
+  or
+  exists(Variable checkedVar, IRGuardCondition guard, Operand access, Operand other |
+    /*
+     * This node is guarded by a condition that forces the accessed variable
+     * to equal something else.  For example:
+     * ```
+     * x = taintsource()
+     * if (x == 10) {
+     *   taintsink(x); // not considered tainted
+     * }
+     * ```
+     */
+
+    readsVariable(node.asInstruction(), checkedVar) and
+    readsVariable(access.getDef(), checkedVar) and
+    guard.ensuresEq(access, other, _, node.asInstruction().getBlock(), true)
```

Comment:
What happens if I compare `x` to another variable that's also tainted? Would it make sense to restrict this case to only block taint when comparing `x` to a constant/predictable instruction?

Reply:
I seem to recall that we've recently had that same discussion in the JavaScript team. @github/codeql-javascript, can you remind me whether we came to any conclusion?

Reply:
Yes, we could go either way on this.  Like much of the security taint tracking library there's a bit of guesswork going on because we don't have all the capabilities of the newer libraries.

My _intuition_ is that most of these comparisons will be the type that makes the value safe, and only a small minority will be with a tainted value.  For example:
```

x = taintsource()
if (x == myarray.size()) { // not a constant
  taintsink(x); // safe
}
```
Also, keeping taint narrow is usually preferred to avoid false positives in queries.  Since this code is blocking taint, we want it to be wide in order to narrow the results that get through and keep false positives to a minimum.  In the case you described where two tainted values are compared, we may lose a true positive result - but we'll hopefully avoid creating many more false positives.

Reply:
Sorry @max-schaefer I didn't see your reply until after I typed mine.  I'm interested to hear what the JavaScript team thinks, though please understand that this is our old taint library and we'd rather put the majority of our effort into the new one.

Reply:
Max recalls https://github.com/github/codeql/pull/3415:

> The most controversial part of this PR is that equality tests are now treated as sanitizers regardless of whether they have a constant side or not.

And https://github.slack.com/archives/CPB7M1KK9/p1588156565270400:

> remind me: why don't we treat if (tainted === something()){ ...} as a sanitizer?

> That's a good question. No reason that I can recall except that the RHS might be user-controlled as well. I don't know how common that is, though.

> we have the (dubious) user-controlled bypass for that problem

---

> Would it make sense to restrict this case to only block taint when comparing x to a constant/predictable instruction?

JS does exactly that. But we are about to change that to permit any RHS. Like @geoffw0  says, this is likely a good trade-off. 
I think this choice is more risky to to in C though since you have so many more dangerous numbers around that may be compared to each other - we pretty much only care about malicious (and rather unique) strings in JavaScript.

Reply:
We only treat comparison with a constant as a sanitizer, but https://github.com/github/codeql/pull/3415 is changing it to be any comparison. Indeed, if both operands are tainted, this can lead to FNs. I'm afraid we don't have any data to back up our decision, so make it what you want.

Reply:
The use of equality as a sanitizer is here being applied in a generic taint-tracking configuration that's shared by 10-20 queries. I think that in this very generic setting we do want to err on the side of fewer alerts. Individual queries can be written with a looser set of sanitizers if it makes sense for their domain.